waku-bindings/src/node/filter.rs

```diff
@@ -4,88 +4,100 @@
 use std::ffi::CString;
 // crates
 use libc::*;
+use std::sync::Arc;
+use tokio::sync::Notify;
 // internal
 use crate::general::contenttopic::WakuContentTopic;
 use crate::general::pubsubtopic::PubsubTopic;
 use crate::general::Result;
 use crate::node::context::WakuNodeContext;
 use crate::utils::{get_trampoline, handle_no_response, LibwakuResponse};
 
-pub fn waku_filter_subscribe(
+pub async fn waku_filter_subscribe(
     ctx: &WakuNodeContext,
     pubsub_topic: &PubsubTopic,
     content_topics: Vec<WakuContentTopic>,
 ) -> Result<()> {
     let content_topics = WakuContentTopic::join_content_topics(content_topics);
 
-    let pubsub_topic_ptr = CString::new(String::from(pubsub_topic))
-        .expect("CString should build properly from pubsub topic")
-        .into_raw();
-    let content_topics_ptr = CString::new(content_topics)
-        .expect("CString should build properly from content topic")
-        .into_raw();
+    let pubsub_topic = CString::new(String::from(pubsub_topic))
+        .expect("CString should build properly from pubsub topic");
+    let pubsub_topic_ptr = pubsub_topic.as_ptr();
+
+    let content_topics =
+        CString::new(content_topics).expect("CString should build properly from content topic");
+    let content_topics_ptr = content_topics.as_ptr();
+
+    let mut result = LibwakuResponse::default();
+    let notify = Arc::new(Notify::new());
+    let notify_clone = notify.clone();
+    let result_cb = |r: LibwakuResponse| {
+        result = r;
+        notify_clone.notify_one(); // Notify that the value has been updated
+    };
```

Comment:
This patterns tends to repeat consistently. I wonder if we can maybe extract this into a function?